tron/serialize/runstate/dynamodb_state_store.py

```diff
@@ -89,7 +112,6 @@ def save(self, key_value_pairs) -> None:
         """
         try:
             for key, val in key_value_pairs:
-                self._delete_item(key)
```

Comment:
I don't think it's safe to assume that the data size will never go down below a partition boundary. How about checking if the number of partitions has changed and only deleting the extra ones if it has gone down, in __setitem__?

Reply:
It's ok if data size goes down. I have a num_partition in the database. So if data size goes down, num_partition is smaller and there would be no errors. The only issue is that the extra data are not deleted and left in the database. But I think that's not a big deal for us.

Reply:
we definitely want to cleanup old data. either leave the _delete_item call or figure out the difference between partition numbers and remove the stale ones.

Reply:
I'm thinking about adding a batch/script to clean them up to make writing faster. We have a limit of 20s right now.

Reply:
How much slower would it be to delete the items?

If deleting is really a blocker, an alternative could be adding a TTL to the items in dynamodb, so they eventually get cleaned up if they haven't been updated in a long time.

Reply:
I believe it could take up to a few seconds. The main reason is that I don't think it's gonna save us much space anyway. Deleting extra partitions here don't really delete removed jobs. I think it's better to delete removed jobs together with a script when DynamoDB is not busy.

Reply:
i'm not a fan of batch script to cleanup those - it's better to be correct at  the spot than relying on somebody cleaning up later.
i also wouldn't worry about 20s window - we only need to flush the buffer during that time, not write all of the state (it's already written). i can't imagine flushing the buffer taking anywhere near 20s (i might be wrong, worth  checking that).

Reply:
That makes sense as long as someone could clean up old jobs. I'll add delete back.

Reply:
yeah, cleanup of old jobs is a separate issue.